src/concurrent.rs

```diff
@@ -1,59 +1,15 @@
-use std::collections::HashMap;
-use std::path::{Path, PathBuf};
-use std::sync::Arc;
-use std::thread;
+use std::path::PathBuf;
 
-use crossbeam::channel::{unbounded, Receiver, Sender};
+use crossbeam::channel::{bounded, Receiver, Sender};
+use crossbeam::scope;
 use globset::GlobSet;
+use rayon::iter::{IntoParallelIterator, ParallelIterator};
 use walkdir::{DirEntry, WalkDir};
 
-use crate::Result;
+use crate::error::Error::Concurrent;
+use crate::{Complexity, Result, Snippets};
 
-type ProcFilesFunction<Config> = dyn Fn(PathBuf, &Config) -> Result<()> + Send + Sync;
-
-type ProcDirPathsFunction<Config> =
-    dyn Fn(&mut HashMap<String, Vec<PathBuf>>, &Path, &Config) + Send + Sync;
-
-type ProcPathFunction<Config> = dyn Fn(&Path, &Config) + Send + Sync;
-
-// Null functions removed at compile time
-fn null_proc_dir_paths<Config>(_: &mut HashMap<String, Vec<PathBuf>>, _: &Path, _: &Config) {}
-fn null_proc_path<Config>(_: &Path, _: &Config) {}
-
-struct JobItem<Config> {
-    path: PathBuf,
-    cfg: Arc<Config>,
-}
-
-type JobReceiver<Config> = Receiver<Option<JobItem<Config>>>;
-type JobSender<Config> = Sender<Option<JobItem<Config>>>;
-
-fn consumer<Config, ProcFiles>(receiver: JobReceiver<Config>, func: Arc<ProcFiles>)
-where
-    ProcFiles: Fn(PathBuf, &Config) -> Result<()> + Send + Sync,
-{
-    while let Ok(job) = receiver.recv() {
-        if job.is_none() {
-            break;
-        }
-        // Cannot panic because of the check immediately above.
-        let job = job.unwrap();
-        let path = job.path.clone();
-
-        if let Err(err) = func(job.path, &job.cfg) {
-            eprintln!("{} for file {:?}", err, path);
-        }
-    }
-}
-
-fn send_file<T>(path: PathBuf, cfg: &Arc<T>, sender: &JobSender<T>) -> Result<()> {
-    sender
-        .send(Some(JobItem {
-            path,
-            cfg: Arc::clone(cfg),
-        }))
-        .map_err(|e| ConcurrentErrors::Sender(e.to_string()).into())
-}
+type ProcFilesFunction = dyn Fn(PathBuf, &[(Complexity, usize)]) -> Option<Snippets> + Send + Sync;
```

Comment:
Is this generic function necessary for `complex-code-spotter`? Can we define a real function and use that instead?  Since we are not exposing any concurrent API as we do for `rust-code-analysis` we can move all the code here in `lib.rs` and simplify the code a bit too